### PR TITLE
feat: Support for map categorical discrete input

### PR DESCRIPTION
**Pull request by nilscb**
### CategoricalPropMap

**After** — categorical property rendering with undefined-value and fallback handling.

### CODE REVIEW

File: typescript/packages/subsurface-viewer/src/layers/map/mapLayer.ts [hunk 1]
```diff
@@ -25,6 +25,13 @@ import { rotate } from "./utils";
 import { makeFullMesh } from "./webworker";
 
 import workerpool from "workerpool";
+import type { RGBColor } from "../../utils";
+
+export type TTypedIntegerArray =
```
Comment (w1nklr) [inline typescript/packages/subsurface-viewer/src/layers/map/mapLayer.ts]: For now only Uint.
We want to ensure positive index to access to the color (otherwise you need to test Int with negative values).

File: typescript/packages/subsurface-viewer/src/layers/grid3d/grid3dLayer.ts [hunk 2]
```diff
@@ -71,7 +71,13 @@ function GetBBox(points: Float32Array): BoundingBox3D {
     return [xmin, ymin, zmin, xmax, ymax, zmax];
 }
 
-type TTypedArray = Float32Array | Uint32Array | Uint16Array;
+type TTypedArray =
```
Comment (w1nklr) [inline typescript/packages/subsurface-viewer/src/layers/grid3d/grid3dLayer.ts:74]: Revert for now.
I'm also realizing that Uint32Array is not supported and could be removed also.

File: typescript/packages/subsurface-viewer/src/layers/map/privateMapLayer.ts [hunk 3]
```diff
@@ -139,7 +179,14 @@ export default class PrivateMapLayer extends Layer<PrivateMapLayerProps> {
...
 
-    _getModels(device: Device) {
+    isPropertiesDiscrete(): boolean {
+        return (
+            this.props.vertexProperties instanceof Uint32Array ||
+            typeof this.props.discretePropertyValueNames !== "undefined"
+        );
+    }
+
+    getLinearModel(device: Device) {
```
Comment (w1nklr) [inline typescript/packages/subsurface-viewer/src/layers/map/privateMapLayer.ts]: Rename to getContinuousPropModel

File: typescript/packages/subsurface-viewer/src/layers/map/privateMapLayer.ts [hunk 4]
```diff
@@ -231,13 +291,124 @@ export default class PrivateMapLayer extends Layer<PrivateMapLayerProps> {
             },
         });
 
-        // MESH LINES
-        const mesh_lines_model = new Model(device, {
+        return model;
+    }
+
+    getCategoricalModel() {
```
Comment (w1nklr) [inline typescript/packages/subsurface-viewer/src/layers/map/privateMapLayer.ts]: rename to getDiscretePropModel

File: typescript/packages/subsurface-viewer/src/layers/map/privateMapLayer.ts [hunk 5]
```diff
@@ -231,13 +291,124 @@ export default class PrivateMapLayer extends Layer<PrivateMapLayerProps> {
...
+                colormapName: this.props.colormapName,
+                colorTables: (this.context as DeckGLLayerContext).userData
+                    .colorTables,
+            },
+            hints
+        );
+
+        const colormapSize = colors.length / 3;
+        const length = this.props.vertexProperties.length;
+        const vertexColors = new Float32Array(length * 3);
```
Comment (w1nklr) [inline typescript/packages/subsurface-viewer/src/layers/map/privateMapLayer.ts]: Why computing the colors here: the goal is to send the property values to the graphics pipeline and let the shader define the color (the same than for continuous property; refer to 3D grid which supports continuous, discrete and categorical properties).

File: typescript/packages/subsurface-viewer/src/layers/map/webworker.ts [hunk 6]
```diff
@@ -38,7 +45,10 @@ export function makeFullMesh(e: { data: Params }) {
         return c as Vec;
     }
 
-    function isDefined(x: unknown): boolean {
+    function isDefined(x: unknown, isDiscrete: boolean = false): boolean {
+        if (isDiscrete) {
+            return typeof x === "number" && x !== 0xffff;
```
Comment (w1nklr) [inline typescript/packages/subsurface-viewer/src/layers/map/webworker.ts]: 0xffff is the default undefined value. You need to check the real undefined value.

File: typescript/packages/subsurface-viewer/src/storybook/layers/MapLayer.stories.tsx [hunk 7]
```diff
@@ -180,6 +180,79 @@ const cellCenteredPropertiesLayer = {
...
+         9,
+         999,
+         11,
+    ],
+
+    discretePropertyValueNames: [
+        { code: 0, name: "Zero" },
+        { code: 1, name: "One" }, 
+        { code: 2, name: "Two" }, 
+        { code: 3, name: "Tree" },
```
Comment (w1nklr) [inline typescript/packages/subsurface-viewer/src/storybook/layers/MapLayer.stories.tsx]: Three

File: typescript/packages/subsurface-viewer/src/layers/map/webworker.ts [hunk 8]
```diff
@@ -184,9 +200,12 @@ export function makeFullMesh(e: { data: Params }) {
         isCellCenteredProperties || !smoothShading ? 0 : nNodes * 3
     );
     const triangleIndices = new Uint32Array(nTriangles * 3);
-    const vertexProperties = new Float32Array(
-        isCellCenteredProperties ? nCells * 6 : nNodes
-    );
+
+    const n = isCellCenteredProperties ? nCells * 6 : nNodes;
+    const vertexProperties = isDiscrete
+        ? new Uint32Array(n)
```
Comment (w1nklr) [inline typescript/packages/subsurface-viewer/src/layers/map/webworker.ts]: Add support for Uint16Array (Viz is sending the property as Uint16Array).

File: typescript/packages/subsurface-viewer/src/layers/map/privateMapLayer.ts [hunk 9]
```diff
@@ -231,13 +291,124 @@ export default class PrivateMapLayer extends Layer<PrivateMapLayerProps> {
...
+        for (let i = 0; i < length; i++) {
+            const property = this.props.vertexProperties[i];
+
+            const index =
+                this.props.discretePropertyValueNames?.findIndex(
```
Comment (w1nklr) [inline typescript/packages/subsurface-viewer/src/layers/map/privateMapLayer.ts]: If discretePropertyValueNames is not specified, all colors are grey !

File: typescript/packages/subsurface-viewer/src/layers/map/privateMapLayer.ts [hunk 10]
```diff
@@ -231,13 +291,124 @@ export default class PrivateMapLayer extends Layer<PrivateMapLayerProps> {
...
+                    (e) => e.code === property
+                ) ?? -1;
+            let color = this.props.discretePropertyValueNames?.[index]?.color; // Use this color if set.
+
+            if (typeof color === "undefined") {
+                if (index !== -1 && property !== 0xffff) {
```
Comment (w1nklr) [inline typescript/packages/subsurface-viewer/src/layers/map/privateMapLayer.ts]: 0xFFFF is the default undefined property. Use the real one.

File: typescript/packages/subsurface-viewer/src/layers/map/privateMapLayer.ts (hunk 5 above)
Comment (w1nklr) [inline typescript/packages/subsurface-viewer/src/layers/map/privateMapLayer.ts]: This should not be computed explicitly on the CPU.
Let the fragment shader do it (see 3D grid cellProperty.fs.glsl.ts).

**Review by w1nklr — DISMISSED** [review]

File: typescript/packages/subsurface-viewer/src/layers/map/mapLayer.ts (hunk 1 above)
Comment (hkfb) [reply typescript/packages/subsurface-viewer/src/layers/map/mapLayer.ts]: Handled. MapLayer now supports only the positive unsigned Uint16Array discrete contract; signed integer inputs are not included.​‌‍​

File: typescript/packages/subsurface-viewer/src/layers/grid3d/grid3dLayer.ts (hunk 2 above)
Comment (hkfb) [reply typescript/packages/subsurface-viewer/src/layers/grid3d/grid3dLayer.ts]: Handled. The added signed typed-array cases were reverted. Grid3D keeps its existing Uint32Array support; MapLayer uses Uint16Array for discrete properties.​‌‍​

File: typescript/packages/subsurface-viewer/src/layers/map/privateMapLayer.ts (hunk 3 above)
Comment (hkfb) [reply typescript/packages/subsurface-viewer/src/layers/map/privateMapLayer.ts]: Renamed to getContinuousPropModel.​‌‍​

File: typescript/packages/subsurface-viewer/src/layers/map/privateMapLayer.ts (hunk 4 above)
Comment (hkfb) [reply typescript/packages/subsurface-viewer/src/layers/map/privateMapLayer.ts]: Renamed to getDiscretePropModel.​‌‍​

File: typescript/packages/subsurface-viewer/src/layers/map/privateMapLayer.ts (hunk 5 above)
Comment (hkfb) [reply typescript/packages/subsurface-viewer/src/layers/map/privateMapLayer.ts]: Deferred for a follow-up as requested: this PR keeps CPU-side color assignment and the existing sampling path.​‌‍​

File: typescript/packages/subsurface-viewer/src/layers/map/webworker.ts (hunk 6 above)
Comment (hkfb) [reply typescript/packages/subsurface-viewer/src/layers/map/webworker.ts]: Handled. The worker now receives and checks the configured undefinedPropertyValue rather than assuming 0xFFFF.​‌‍​

File: typescript/packages/subsurface-viewer/src/storybook/layers/MapLayer.stories.tsx (hunk 7 above)
Comment (hkfb) [reply typescript/packages/subsurface-viewer/src/storybook/layers/MapLayer.stories.tsx]: Fixed to Three.​‌‍​

File: typescript/packages/subsurface-viewer/src/layers/map/webworker.ts (hunk 8 above)
Comment (hkfb) [reply typescript/packages/subsurface-viewer/src/layers/map/webworker.ts]: Added Uint16Array support end to end, including loading, worker range handling, and rendering.​‌‍​

File: typescript/packages/subsurface-viewer/src/layers/map/privateMapLayer.ts (hunk 9 above)
Comment (hkfb) [reply typescript/packages/subsurface-viewer/src/layers/map/privateMapLayer.ts]: Handled. Without discretePropertyValueNames, non-negative integer property codes now index the generated colormap and fall back safely.​‌‍​

File: typescript/packages/subsurface-viewer/src/layers/map/privateMapLayer.ts (hunk 10 above)
Comment (hkfb) [reply typescript/packages/subsurface-viewer/src/layers/map/privateMapLayer.ts]: Handled. The configured undefined value is propagated to rendering and used for the undefined color.​‌‍​

File: typescript/packages/subsurface-viewer/src/layers/map/privateMapLayer.ts (hunk 5 above)
Comment (hkfb) [reply typescript/packages/subsurface-viewer/src/layers/map/privateMapLayer.ts]: Deferred for a follow-up as requested: shader-side color selection is intentionally unchanged in this PR.​‌‍​

File: typescript/packages/subsurface-viewer/src/layers/map/privateMapLayer.ts [hunk 11]
```diff
@@ -100,6 +115,31 @@ export default class PrivateMapLayer extends Layer<PrivateMapLayerProps> {
...
+            this.props.vertexProperties instanceof Int32Array ||
+            this.props.vertexProperties instanceof Uint16Array ||
+            this.props.vertexProperties instanceof Int16Array ||
+            typeof this.props.discretePropertyValueNames !== "undefined"
+        ) {
+            return {
+                discreteData: true,
+                colormapSize: (this.props.propertyValueRange?.[1] ?? 0) + 1,
+            };
+        }
```
[the remark this answers is not in the document]
Comment (hkfb) [reply typescript/packages/subsurface-viewer/src/layers/map/privateMapLayer.ts:143]: Handled for named categories: colormap width now uses the number of named entries; unnamed values use the CPU-side fallback path. Shader-side selection remains deferred.​‌‍​

File: typescript/packages/subsurface-viewer/src/layers/map/privateMapLayer.ts [hunk 12]
```diff
@@ -139,7 +179,14 @@ export default class PrivateMapLayer extends Layer<PrivateMapLayerProps> {
         this.initializeState(context as DeckGLLayerContext);
     }
 
-    _getModels(device: Device) {
+    isPropertiesDiscrete(): boolean {
+        return (
+            this.props.vertexProperties instanceof Uint32Array ||
+            typeof this.props.discretePropertyValueNames !== "undefined"
+        );
+    }
```
[the remark this answers is not in the document]
Comment (hkfb) [reply typescript/packages/subsurface-viewer/src/layers/map/privateMapLayer.ts]: Handled by aligning detection with the supported contract: Uint16Array and named categorical inputs select discrete mode; unsupported signed and Uint32Array MapLayer paths were removed.​‌‍​

File: typescript/packages/subsurface-viewer/src/layers/map/privateMapLayer.ts [hunk 13]
```diff
@@ -231,13 +291,124 @@ export default class PrivateMapLayer extends Layer<PrivateMapLayerProps> {
...
+        const vertexColors = new Float32Array(length * 3);
+        for (let i = 0; i < length; i++) {
+            const property = this.props.vertexProperties[i];
+
+            const index =
+                this.props.discretePropertyValueNames?.findIndex(
+                    (e) => e.code === property
+                ) ?? -1;
+            let color = this.props.discretePropertyValueNames?.[index]?.color; // Use this color if set.
+
```
[the remark this answers is not in the document]
Comment (hkfb) [reply typescript/packages/subsurface-viewer/src/layers/map/privateMapLayer.ts]: Handled with a precomputed Map<number, number> for O(1) per-vertex lookups.​‌‍​

File: typescript/packages/subsurface-viewer/src/layers/map/webworker.ts [hunk 14]
```diff
@@ -19,10 +20,16 @@ export function makeFullMesh(e: { data: Params }) {
     // Local functions.
     type Vec = [number, number, number];
 
-    function getFloat32ArrayMinMax(data: Float32Array) {
+    function getArrayMinMax(
+        data: TypedArray,
+        isDiscrete: boolean = false
+    ): [number, number] {
         let max = -99999999;
```
[the remark this answers is not in the document]
Comment (hkfb) [reply typescript/packages/subsurface-viewer/src/layers/map/webworker.ts]: Handled by using infinity initialization and returning [0, 0] when no defined values are present.​‌‍​

File: typescript/packages/subsurface-viewer/src/layers/map/webworker.ts [hunk 15]
```diff
@@ -38,7 +45,10 @@ export function makeFullMesh(e: { data: Params }) {
         return c as Vec;
     }
 
-    function isDefined(x: unknown): boolean {
+    function isDefined(x: unknown, isDiscrete: boolean = false): boolean {
+        if (isDiscrete) {
+            return typeof x === "number" && x !== 0xffff;
+        }
         return typeof x === "number" && !isNaN(x);
     }
```
[the remark this answers is not in the document]
Comment (hkfb) [reply typescript/packages/subsurface-viewer/src/layers/map/webworker.ts]: Handled by limiting discrete MapLayer inputs to Uint16Array; signed arrays are no longer reinterpreted.​‌‍​

File: typescript/packages/subsurface-viewer/src/layers/map/webworker.ts [hunk 16]
```diff
@@ -184,9 +200,12 @@ export function makeFullMesh(e: { data: Params }) {
...
     );
     const triangleIndices = new Uint32Array(nTriangles * 3);
-    const vertexProperties = new Float32Array(
-        isCellCenteredProperties ? nCells * 6 : nNodes
-    );
+
+    const n = isCellCenteredProperties ? nCells * 6 : nNodes;
+    const vertexProperties = isDiscrete
+        ? new Uint32Array(n)
+        : new Float32Array(n);
```
[the remark this answers is not in the document]
Comment (hkfb) [reply typescript/packages/subsurface-viewer/src/layers/map/webworker.ts]: Handled: discrete vertex properties are allocated as Uint16Array.​‌‍​

File: typescript/packages/subsurface-viewer/src/layers/utils/colormapTools.ts [hunk 17]
```diff
@@ -199,3 +200,27 @@ export function createColormapTexture(
...
 }
+
+export function getColormapDiscreteColors(
+    colormap: ColormapProps,
+    colormapHints: IColormapHints
+): Uint8Array {
+    if (colormapHints.discreteData) {
+        if (colormapHints.colormapSize === 0) {
+            return new Uint8Array([0, 0, 0, 0, 0, 0]);
+        }
```
[the remark this answers is not in the document]
Comment (hkfb) [reply typescript/packages/subsurface-viewer/src/layers/utils/colormapTools.ts]: Handled: empty discrete colormaps now produce one fallback color and a minimum texture width of one.​‌‍​

File: typescript/packages/subsurface-viewer/src/layers/grid3d/grid3dLayer.ts [hunk 18]
```diff
@@ -71,7 +71,13 @@ function GetBBox(points: Float32Array): BoundingBox3D {
...
+    | Float32Array
+    | Uint16Array
+    | Int32Array
+    | Uint32Array
+    | Uint16Array
+    | Int16Array;
```
[the remark this answers is not in the document]
Comment (hkfb) [reply typescript/packages/subsurface-viewer/src/layers/grid3d/grid3dLayer.ts]: Removed the duplicate Uint16Array entry from the Grid3D typed-array union.​‌‍​

File: typescript/packages/subsurface-viewer/src/layers/map/mapLayer.ts [hunk 19]
```diff
@@ -128,9 +162,28 @@ export interface MapLayerProps extends ExtendedLayerProps {
      * If the number of property values equals one less than the depth values in
      * each direction then the property values will be pr cell and the cell will be constant
      * colored.
+     * Undefined value for discrete (Uint32Array) input is 0xFFFF. For float input (Float32Array) it is NaN.
+     * For number array use "undefined".
      */
```
[the remark this answers is not in the document]
Comment (hkfb) [reply typescript/packages/subsurface-viewer/src/layers/map/mapLayer.ts]: Handled: discrete number[] inputs normalize undefined to the configured sentinel without mutating caller data, and the docs now describe this behavior.​‌‍​

File: typescript/packages/subsurface-viewer/src/storybook/layers/MapLayer.stories.tsx [hunk 20]
```diff
@@ -180,6 +180,79 @@ const cellCenteredPropertiesLayer = {
     smoothShading: true,
 };
 
+// NB maa vaere like mange som ..??
```
[the remark this answers is not in the document]
Comment (hkfb) [reply typescript/packages/subsurface-viewer/src/storybook/layers/MapLayer.stories.tsx]: Reworded the Storybook comments in clear English.​‌‍​

File: typescript/packages/subsurface-viewer/src/storybook/layers/MapLayer.stories.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (hkfb) [reply typescript/packages/subsurface-viewer/src/storybook/layers/MapLayer.stories.tsx]: Fixed Tree to Three.​‌‍​

File: typescript/packages/subsurface-viewer/src/storybook/layers/MapLayer.stories.tsx [hunk 21]
```diff
@@ -180,6 +180,79 @@ const cellCenteredPropertiesLayer = {
...
+         11,
+    ],
+
+    discretePropertyValueNames: [
+        { code: 0, name: "Zero" },
+        { code: 1, name: "One" }, 
+        { code: 2, name: "Two" }, 
+        { code: 3, name: "Tree" }, 
+        { code: 4, name: "Four" }, 
+        { code: 5, name: "Five", color: [255, 0, 255] }, // explisit color overrides colortable.
```
[the remark this answers is not in the document]
Comment (hkfb) [reply typescript/packages/subsurface-viewer/src/storybook/layers/MapLayer.stories.tsx]: Fixed explisit to explicit.​‌‍​

**Review by hkfb — APPROVED** [review]
